Add forked repositories

padraic-padraic opened the pull request:
This adds support for creating a repository using a fork. This was implemented on our fork (as documented [here](https://registry.terraform.io/providers/Zensor-BE/bitbucket/latest/docs/resources/repository)), but had to been reworked to use the API client module.

I haven't had a chance to test this yet, just opening the PR for feedback!

padraic-padraic commented:
For reference, here's the old PR on the previous provider: https://github.com/aeirola/terraform-provider-bitbucket/pull/6

DrFaust92 commented:
Hi [padraic-padraic](https://github.com/padraic-padraic), thanks for adding this. 
can you add some acceptance tests as well? Ill try to add/test this if need me to.

DrFaust92 commented:
Hi [padraic-padraic](https://github.com/padraic-padraic), i have tested this on the side and this code doesnt compile. i tried fixing it up a bit but it fails to fork the repo. please take a look.

padraic-padraic commented:
Sure thing,  working on it now. Apologies, not much of a go dev so I had thought passing fmt also meant it would compile.

Found the issue in the API causing the read step to crash, testing now and will push it tonight.

padraic-padraic commented:
I've managed to fix creating repositories too!

Some quirks that I ran into the first time I implemented this. Firstly, for whatever reason, calling the pipelines config API with a forked repo is very fussy about the slug; it gives a 404 if provided the repo name instead of the slug. So I added a method for computing slugs, and additionally added a DiffSupressFunc to stop the provider from trying to correct e.g. capitalised slugs.

The other quirk is that it can take a few for applying the pipeline config on a forked repo to work; initially it gives a 403 error. So I added a retry block which will retry applying the pipeline config.

Will work on adding some acceptance tests now!

DrFaust92 commented on bitbucket/resource_repository.go:
```diff
@@ -124,6 +132,27 @@ func resourceRepository() *schema.Resource {
+				},
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
```
why? default is 10 mins why would it take up to an hour to make an API call?

padraic-padraic replied:
Indeed, this can be stripped out; this is a legacy from a prior implementation, before I realised the cause of the API errors. I did think it was setup for 60s rather than 60 minutes.

DrFaust92 commented on bitbucket/resource_repository.go:
```diff
@@ -154,6 +183,46 @@ func newRepositoryFromResource(d *schema.ResourceData) *bitbucket.Repository {
+	Slug string `json:"slug,omitempty"`
+}
+
+type forkedRepositoryBody struct {
+	Name string `json:"name,omitempty"`
+	Language string `json:"language,omitempty"`
+	IsPrivate bool `json:"is_private,omitempty"`
+	Description string `json:"description,omitempty"`
+	ForkPolicy string `json:"fork_policy,omitempty"`
+	HasWiki bool `json:"has_wiki,omitempty"`
+	HasIssues bool `json:"has_issues,omitempty"`
+	Links *bitbucket.RepositoryLinks `json:"links,omitempty"`
+	Project *bitbucket.Project `json:"project,omitempty"`
+	Workspace *forkWorkspace `json:"workspace,omitempty"`
+}
+
+func createForkedRepositoryFromRepository(repo *bitbucket.Repository, targetWorkspaceSlug string) *forkedRepositoryBody{
+	forkedRepo := &forkedRepositoryBody{
+		Name: repo.Name,
+		Language: repo.Language,
+		IsPrivate: repo.IsPrivate,
+		Description: repo.Description,
+		ForkPolicy: repo.ForkPolicy,
+		HasWiki: repo.HasWiki,
+		HasIssues: repo.HasIssues,
+	}
+	workspace := &forkWorkspace{
```
why not reuse Repository struct? why is it needed to create another one

padraic-padraic replied:
The fork requires an additional field, workspace; unfortunately this doesn't seem to be documented in the OpenAPI spec (or rather, it is; "Extra properties of any type may be provided to this object.", to quote the API docs), in a way that the OpenAPI spec doesn't quite capture.

DrFaust92 reviewed (changes requested):
A few comments, plus:

* Add acceptance tests + run existing ones to ensure no regressions fails happen
* Fix linting errors
* Strongly consider a dedicate forked_repository resource instead as this really clusters the repository resource.

padraic-padraic commented:
Happy to create a separate resource if that would be preferable; the update and delete logic are identical but indeed the create step gets quite complicated with the need for the retryable step and the additional struct.

If I split up the code, are you happy to leave the changes regarding the slug in `bitbucket/resource_repository.go`?

DrFaust92 commented:
[padraic-padraic](https://github.com/padraic-padraic), as long as with slug changes all tests pass, i dont mind.

but i think its best to have a dedicated resource for fork as the api to create is different. also you can reuse update/delete in the the fork resource definition.

padraic-padraic commented:
Spliting the resources is done, will run acceptance tests shortly but my own testing shows the resource working as expected.

DrFaust92 commented:
@padraic-padraic linter fails but otherwise looks good. ill try to run acctests ASAP to confirm changes

DrFaust92 reviewed (approved):
LGTM

DrFaust92 commented:
@padraic-padraic missing acceptance tests for forked repos? even something basic will suffice (create/update)

DrFaust92 commented:
+ Spelling error on linter

giveadamakick commented:
@DrFaust92 @padraic-padraic I would love to see this feature merged and released. Is there anything I can do to help?

padraic-padraic commented:
Apologies, haven't had a lot of spare time lately (I just became a dad), but will work on adding those integration tests ASAP

DrFaust92 commented:
Thanks [padraic-padraic](https://github.com/padraic-padraic), added tests and fixed lint issues.
